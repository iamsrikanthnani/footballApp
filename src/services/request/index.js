import request from './request.service';

export default request;
export * from './request.service';

export {
  default as responseHandler,
  responseHandlerWithoutStatusCheck,
  responseHandlerGraphQL,
} from './responseHandler/responseHandler';
