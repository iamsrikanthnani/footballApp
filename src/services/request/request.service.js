import omit from 'lodash/omit';
// import Base64 from 'Base64';
import get from 'lodash/get';
import { Platform } from 'react-native';
import Config from 'react-native-config';
// import { strings } from '../localization';
import ResponseError from './responseError';
// import { bugsnag } from '../analytics/bugsnag.service';
import { underScoreToCamel } from '../../utilities/case';

/*
 * Since this file is the only place we should be using fetch, tell eslint
 * about the fetch global here but let it yell at us for using it in any other
 * place.
 */
/* global fetch */

/* A Function that will handling building and firing off a http request.
 * @param {Object} requestOptions - Used to build the request.
 * @return {Promise}
 */
/*
 * requestOptions Fields:
 * @param {String} url - *Required* the url to be used for the request
 * @param {String} method - *Required* the http method to use
 * @param {Object} headers - any additional items that need to be added to the header
 */
export default function request(requestOptions) {
  if (!requestOptions.method || !requestOptions.url) {
    throw new Error('RequestOptions missing required property method or url');
  }

  const options = omit(requestOptions, ['headers', 'userCredentials', 'url']);
  return fetch(requestOptions.url, { ...options, headers: combineHeaders(requestOptions) })
    .then(response => {
      if (!response.ok) {
        const error = new ResponseError(get(response, 'status'));
        throwGenericError(error);
        // reportError(error, requestOptions);
      }
      return response;
    })
    .then(response => {
      switch (response.status) {
        case 204:
          return null;
        case 202:
          return throwGenericError(new ResponseError(get(response, 'status')), requestOptions);
        default:
          return response.json();
      }
    })
    .catch(error => throwGenericError(error, requestOptions));
}

function throwGenericError(error, requestOptions) {
  // reportError(error, requestOptions);
  throw new ResponseError(get(error, 'status'), 'strings(App.genericServiceError)');
}

/* TODO: remove in favor of request/errorForStatus
 * @param {String} status - The status received from the service
 * @param {String} localizationPath - The path to the list of errors in the language files.
 */
function errorForStatus(status, localizationPath) {
  let errorMessage = 'strings(App.genericError)';
  if (status) {
    const underScoreStatus = underScoreToCamel(status);
    errorMessage = 'StringHolder';
    // strings(`${localizationPath}.${underScoreStatus}`, { defaultValue: errorMessage });
  }
  return errorMessage;
}

function setDefaultHeaders(headers) {
  const { APP_NAME } = Config;
  const { graphQL } = headers;

  let allHeaders = {
    'Accept-Encoding': 'gzip,deflate',
    'Accept-Language': 'en-US',
    'Content-Type': graphQL ? 'text/plain' : 'application/json',
    'User-Agent': `${APP_NAME} (compatible; Mozilla/5.0; MSIE 9.0; Trident/5.0; ${Platform.OS})`,
    'Proxy-Connection': 'Keep-Alive',
    Connection: 'Keep-Alive',
    Accept: 'application/hal+json, application/json',
    ...headers,
  };

  // Omit Accept-Encoding header for Android only since it will gzip encoding on it's own
  if (Platform.OS === 'android') {
    allHeaders = omit(allHeaders, 'Accept-Encoding');
  }

  return allHeaders;
}

function combineHeaders(requestOptions) {
  // TODO: There are other items that will eventually need to be added to the header, like impersonation items.
  const optionHeaders = requestOptions.headers || {};
  if (requestOptions.userCredentials) {
    return setDefaultHeaders({ ...optionHeaders, ...getCredentials(requestOptions.userCredentials) });
  }
  return setDefaultHeaders(optionHeaders);
}

function getCredentials({ username, password }) {
  const base = { username, password }
  // `Basic ${Base64.btoa(`${username}:${password}`)}`;
  return { Authorization: base };
}

// https://docs.bugsnag.com/platforms/react-native/#attaching-custom-diagnostics
/* eslint-disable no-param-reassign, prefer-arrow-callback, func-names */
// function reportError(error, requestOptions, response) {
//   bugsnag.notify(error, function (report) {
//     report.metadata = {
//       request: {
//         url: requestOptions ? requestOptions.url : null,
//         method: requestOptions ? requestOptions.method : null,
//       },
//       response: {
//         status: response ? response.status : null,
//       },
//     };
//   });
// }
export { getCredentials, setDefaultHeaders, combineHeaders, errorForStatus };
