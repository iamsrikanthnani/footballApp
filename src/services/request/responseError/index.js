import * as _ from 'lodash';

function ResponseError(status, message) {
  const instance = new Error(message);
  instance.name = 'ResponseError';
  instance.status = status;

  Object.setPrototypeOf(instance, Object.getPrototypeOf(this));
  return instance;
}

ResponseError.prototype = Object.create(Error.prototype, {
  constructor: {
    value: Error,
    enumerable: false,
    writable: true,
    configurable: true,
  },
});

if (_.isFunction(Object.setPrototypeOf)) {
  Object.setPrototypeOf(ResponseError, Error);
} else {
  // setting __proto__ manually like this is deprecated, but is required for android
  ResponseError.__proto__ = Error; // eslint-disable-line no-proto
}

export default ResponseError;
