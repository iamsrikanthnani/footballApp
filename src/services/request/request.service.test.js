import request, { combineHeaders } from '.';
import Methods from '../../utilities/methods';

const requestOptions = {
  url: 'http://example.com',
  method: Methods.GET,
  body: {
    bacon: 'baconbaconbacon',
  },
  headers: {
    BreakfastFoods: 'AreTheBest',
  },
};

const requestOptionsWithCredentials = {
  ...requestOptions,
  userCredentials: { username: 'demo', password: 'demo' },
};

jest.mock('Platform', () => {
  const Platform = require.requireActual('Platform');
  Platform.OS = 'ios';

  return Platform;
});
describe('Request Service Header Handling', () => {
  it('combineHeaders() can combine optional and default headers', () => {
    const result = combineHeaders(requestOptions);
    expect(result).toMatchSnapshot();
  });

  it('combineHeaders() can add basic authentication when necessary', () => {
    const resultWithoutBasicAuth = combineHeaders(requestOptions);
    const resultWithBasicAuth = combineHeaders(requestOptionsWithCredentials);

    expect(resultWithBasicAuth).not.toEqual(resultWithoutBasicAuth);
    expect(resultWithBasicAuth).toMatchSnapshot();
  });
});

describe('Request Service Request Handling', () => {
  it('request() can make an HTTP request using fetch', async () => {
    const json = jest.fn().mockImplementation(options => Promise.resolve({ ...options }));
    const res = options => ({ ok: true, json: () => json(options) });
    global.fetch = jest.fn().mockImplementation(options => Promise.resolve(res(options)));

    const response = await request(requestOptions);
    expect(response).toMatchSnapshot();
  });

  it('request() can throw when missing required parameters', () => {
    expect(() => request({})).toThrow('RequestOptions missing required property method or url');
  });

  it('request() can return an error message when an http request fails', () => {
    global.fetch = jest.fn().mockImplementation(() => Promise.reject());

    return request({ url: 'http://example.com', method: Methods.POST })
      .catch(err => expect(err).toMatchSnapshot());
  });

  it('request() can throw an error when the response is not ok', async () => {
    const json = jest.fn().mockImplementation(options => Promise.resolve({ ...options }));
    const res = options => ({ ok: false, json: () => json(options) });
    global.fetch = jest.fn().mockImplementation(options => Promise.resolve(res(options)));

    return request(requestOptions)
      .catch(err => expect(err).toMatchSnapshot());
  });

  it('request() can throw an error when the response status is 202', async () => {
    const json = jest.fn().mockImplementation(options => Promise.resolve({ ...options }));
    const res = options => ({ status: 202, json: () => json(options) });
    global.fetch = jest.fn().mockImplementation(options => Promise.resolve(res(options)));

    return request(requestOptions)
      .catch(err => expect(err).toMatchSnapshot());
  });
});
