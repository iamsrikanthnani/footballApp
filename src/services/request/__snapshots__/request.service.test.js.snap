// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Request Service Header Handling combineHeaders() can add basic authentication when necessary 1`] = `
Object {
  "Accept": "application/hal+json, application/json",
  "Accept-Encoding": "gzip,deflate",
  "Accept-Language": "en-US",
  "Authorization": "Basic ZGVtbzpkZW1v",
  "BreakfastFoods": "AreTheBest",
  "Connection": "Keep-Alive",
  "Content-Type": "application/json",
  "Proxy-Connection": "Keep-Alive",
  "User-Agent": "carrier360 (compatible; Mozilla/5.0; MSIE 9.0; Trident/5.0; ios)",
}
`;

exports[`Request Service Header Handling combineHeaders() can combine optional and default headers 1`] = `
Object {
  "Accept": "application/hal+json, application/json",
  "Accept-Encoding": "gzip,deflate",
  "Accept-Language": "en-US",
  "BreakfastFoods": "AreTheBest",
  "Connection": "Keep-Alive",
  "Content-Type": "application/json",
  "Proxy-Connection": "Keep-Alive",
  "User-Agent": "carrier360 (compatible; Mozilla/5.0; MSIE 9.0; Trident/5.0; ios)",
}
`;

exports[`Request Service Request Handling request() can make an HTTP request using fetch 1`] = `
Object {
  "0": "h",
  "1": "t",
  "10": "m",
  "11": "p",
  "12": "l",
  "13": "e",
  "14": ".",
  "15": "c",
  "16": "o",
  "17": "m",
  "2": "t",
  "3": "p",
  "4": ":",
  "5": "/",
  "6": "/",
  "7": "e",
  "8": "x",
  "9": "a",
}
`;

exports[`Request Service Request Handling request() can return an error message when an http request fails 1`] = `[ResponseError: Service temporarily unavailable]`;

exports[`Request Service Request Handling request() can throw an error when the response is not ok 1`] = `[ResponseError: Service temporarily unavailable]`;

exports[`Request Service Request Handling request() can throw an error when the response status is 202 1`] = `[ResponseError: Service temporarily unavailable]`;
